src/ncp/spinel.c

```diff
@@ -332,7 +331,14 @@ spinel_datatype_vunpack_(const uint8_t *data_ptr, spinel_size_t data_len, const
 
             if (arg_ptr)
             {
-                *arg_ptr = (spinel_ipv6addr_t *)data_ptr;
+                if (*arg_ptr)
+                {
+                    memcpy(*arg_ptr, data_ptr, sizeof(spinel_ipv6addr_t));
+                }
+                else
+                {
+                    *arg_ptr = (spinel_ipv6addr_t *)data_ptr;
+                }
```

Comment:
@bukepo I understand the optimization here and I agree it will helpful in some case, but I think it's not a good idea to change the behavior of the spinel C APIs for two reasons:

1) The `spinel.c` and `spinel.h` source files are directly integrated and used in other projects (e.g., in `wpantund` and some other internal projects). Changing this will have an impact on other projects.

2) While this makes the API more powerful, at the same time it becomes very error-prone and sort of dangerous. If the user/caller uses a local pointer variable (which is a very common use pattern) and forgets to initialize it to `NULL` then the data can be copied into some unknown/random memory region leading to memory corruption.  

I think it'd be safer to keep the C spinel APIs simple and safe and not change them.

I have been working on adding C++ wrapper classes over spinel C APIs (a `SpinelEncoder` and `SpinelDecoder`) to add more type-checking/safety to the code and help simplify `NcpBase` getter/setter methods. It's still work under progress (the `SpinelEncoder` code is in my OT fork in this (under-dev) branch https://github.com/abtink/openthread/commits/pr/spinel-encoder). We can consider adding support for such behavior in the C++ wrapper classes.